ted2zim/multi/scraper.py

```diff
@@ -0,0 +1,330 @@
+            logger.error(f"{key} failed to upload to cache: {exc}")
+        else:
+            logger.info(f"successfully uploaded playlist list to cache at {key}")
+        finally:
+            fpath.unlink()
+
+    def get_list_of_all(self, mode):
+        """ returns a list of topics or playlists"""
+        if mode != "topic" and mode != "playlist":
+            raise ValueError(f"Mode {mode} not supported. Use playlist or topic")
+        # get all topics
+        topic_list = json.loads(
+            download_link("https://www.ted.com/topics/combo?models=Talks").text
+        )
+        if mode == "topic":
+            return topic_list
+
+        # mode is playlist
+        s3_url_with_credentials = None
+        for arg in self.extra_args:
+            if "--optimization-cache" in arg:
+                # handle both = seperated and space seperated values
+                if len(arg) > 21:
+                    s3_url_with_credentials = arg[21:]
+                else:
+                    s3_url_with_credentials = self.extra_args[
+                        self.extra_args.index(arg) + 1
+                    ]
+
+        if s3_url_with_credentials:
+            s3_storage = KiwixStorage(s3_url_with_credentials)
+            if not s3_storage.check_credentials(
+                list_buckets=True, bucket=True, write=True, read=True, failsafe=True
+            ):
+                logger.error("S3 credential check failed. Continuing without S3")
+                return self.download_playlist_list_from_site(topic_list)
+            self.output_dir.mkdir(parents=True, exist_ok=True)
+            playlist_list = self.download_playlist_list_from_cache(s3_storage)
+            if not playlist_list:
+                playlist_list = self.download_playlist_list_from_site(topic_list)
+                self.upload_playlist_list_to_cache(playlist_list, s3_storage)
+            return playlist_list
+        return self.download_playlist_list_from_site(topic_list)
+
+    def run(self):
+        def log_run_result(success, process):
+            if success:
+                logger.info(".. OK")
+            else:
+                logger.error(".. ERROR. Printing scraper output and exiting.")
+                logger.error(process.stdout)
+                return process.returncode
+
+        logger.info(f"starting {NAME}-multi scraper")
+
+        self.fetch_metadata()
+
+        if self.topics:
+            if self.topics == ["all"]:
+                self.topics = [
+                    topic["value"] for topic in self.get_list_of_all(mode="topic")
+                ]
+            if self.indiv_zims:
+                for topic in self.topics:
+                    logger.info(f"Executing ted2zim for topic {topic}")
+                    success, process = self.run_indiv_zim_mode(topic, mode="topic")
+                    log_run_result(success, process)
+
+            else:
+                logger.info("Falling back to ted2zim for topic(s)")
```

Comment:
Do we want to log that?

Reply:
I get it. It's unnecessary as ted2zim logs that it has started.